### ISSUE TITLE
Method type argument inference: lambda parameter types and method references

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1384]
Taking current master, including #1380 and #1381 but they shouldn't impact this, run the Nullness Checker on:

```java
interface Box<T> {}

interface Function<P, R> {
    R apply(P p);
}

interface Utils {
  <I, O> Box<O> foo(
      Box<I> input, Function<? super I, ? extends O> function);
  <I, O> Function<I, O> bar(
      Function<? super I, ? extends O> function);
}

class FalsePositive {
  Box<Integer> demo(Utils u, Box<String> bs) {
    return u.foo(
        bs,
        u.bar(
            (String s) -> 5
              )
                 );
  }
}
```

One gets:

```
FalsePositives.java:19: error: [lambda.param.type.incompatible] incompatible parameter types in lambda expression.
            (String s) -> 5
                    ^
  found   : ? extends @Initialized @Nullable Object
  required: @Initialized @NonNull String
1 error
```

The uninferred wildcard should be ignored (if not in conservative mode).

**Comment by wmdietlGC** [comment]
Using a method reference results in the same problem. Full code below (could be combined with code above):

```java
interface Box<T> {}

interface Function<P, R> {
    R apply(P p);
}

interface Utils {
    <I, O> Box<O> foo(Box<I> input, Function<? super I, ? extends O> function);

    <I, O> Function<I, O> bar(Function<? super I, ? extends O> function);
}

class FalsePositive {
    Integer ugh(String n) { return 5; }

    Box<Integer> demo(Utils u, Box<String> bs) {
        return u.foo(bs, u.bar(this::ugh));
    }
}
```

Running the Nullness Checker gives:

```
FalsePositives.java:17: error: [methodref.param.invalid] Incompatible parameter type
        return u.foo(bs, u.bar(this::ugh));
                               ^
  Method
    @Initialized @NonNull Integer ugh(@Initialized @NonNull FalsePositive this, @Initialized @NonNull String p0) in FalsePositive
    is not a valid method reference for
    @Initialized @NonNull Object apply(@Initialized @NonNull Function<? extends @Initialized @Nullable Object, @Initialized @NonNull Object> this, ? extends @Initialized @Nullable Object p0) in Function
  found   : @Initialized @NonNull String
  required: ? extends @Initialized @Nullable Object
1 error
```

**Comment by smillst** [comment]
Fixed in #1387. (Both examples.)
